L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,259 @@
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
+* Resolver will push addresses to `ConnectionManager` which will then call balancer.
+* Balancer will use `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
+  * A new sub-channel will start connecting when `SubChannel.ConnectAsync` is called. A sub-channel will be shutdown when disposed.
+* Balancer will be notified on changes in sub-channel state. It will create a `SubChannelPicker` based on the current state.
+
+### Picking
+
+* `ConnectionManager.PickAsync` is called to get a ready sub-channel along with the sub-channel's connected address.
+* If a picker hasn't been set yet, or the picker is empty, or the picker has an error then `PickAsync` will wait for a picker with a ready sub-channel if `WaitForReady` has been set. If `WaitForReady` is not set then the error will be thrown back to the caller.
+
+## Rationale
+
+* Design shares the most in common with grpc-go. SubChannels have a collection of addresses and are responsible for staying connected and retrying when in TRANSIENT_FAILURE.
+* Goal is to support basic gRPC load balancing using DNS in Kubernetes.
+* Secondary goal is to have a design in place that can be used to implement xDS in the future with minimal changes.
+* All load balancing APIs will be marked as experimental.
+* API design of resolver and load balancer purposefully choses not to use static configuration. This is considered an anti-pattern in .NET. Instead custom resolver and load balancer factories are registered with a DI container which is then hooked up to the gRPC channel using `GrpcChannelOptions.ServiceProvider`.
+
+## Implementation
+
+Implementation is in the grpc-dotnet repo: [PR #1286](https://github.com/grpc/grpc-dotnet/pull/1286).
+
+```csharp
+namespace Grpc.Net.Client.Balancer {
+    public sealed class AddressResolverOptions {
+        public AddressResolverOptions();
+    }
+    public class SubChannelOptions {
+        public SubChannelOptions(IReadOnlyList<DnsEndPoint> addresses);
+        public IReadOnlyList<DnsEndPoint> Addresses { get; }
+    }
+    public interface IConnectionController {
+        ConnectivityState State { get; }
+        SubChannel CreateSubChannel(SubChannelOptions options);
+        void RemoveSubChannel(SubChannel subChannel);
+        Task ResolveNowAsync(CancellationToken cancellationToken);
+        void UpdateAddresses(SubChannel subChannel, IReadOnlyList<DnsEndPoint> addresses);
+        void UpdateState(BalancerState state);
+    }
+    public abstract class AddressResolver : IObservable<AddressResolverResult>, IDisposable {
+        protected AddressResolver();
+        public void Dispose();
+        public abstract Task RefreshAsync(CancellationToken cancellationToken);
```

Comment:
This method is async and it is best practice to pass in a cancellation token on async methods.

In this case, the cancellation token will belong to the channel. If the overall channel is shut down then the cancellation token would trigger, alerting this method that it should stop what it is doing.

Reply:
How is this different from a non-refresh resolution, which the resolver does automatically when it is started?  There is no cancellation token in that case, and the channel can still shut down the resolver, which should result in any currently in-flight resolution request being cancelled.

Also, consider the case of a watch-based resolver, where there is essentially *always* a resolution request ongoing from the moment the resolver is started until the moment it is shut down.  In that case, `RefreshAsync()` is a no-op, but the resolver can still be shut down.

I basically don't see why the channel needs to know or care whether there is a refresh request going on within the resolver at any given time.  It already needs the ability to shut down the resolver without knowing whether or not there is currently a request in flight; the latter is the internal business of the resolver implementation.

To say this another way, if there is a cancellation token needed for the underlying request, then I think this is something that the resolver implementation needs to track itself.  It would pass the cancellation token to the underlying request when it gets shut down by the channel.

Reply:
> How is this different from a non-refresh resolution, which the resolver does automatically when it is started? There is no cancellation token in that case, and the channel can still shut down the resolver, which should result in any currently in-flight resolution request being cancelled.

`RefreshResolver` is a public method and is called from user code. The call eventually makes its way to `Resolver.RefreshAsync`, but when that happens the async method is passed the cancellation token for the overall channel. User code will never call `Resolver.RefreshAsync` directly.

Yes, the cancellation token could be removed from this method. But it would force each resolver implementation to:
* Allocate its own cancellation token source
* Ensure that all async methods in RefreshAsync use it
* Trigger the cancellation token source when disposed

On the other hand, passing in a cancellation token to an async method is:
* A standard pattern in .NET
* Allocates less memory
* Removes complexity from resolver implementations, and
* Tooling will alert developers if they call async methods of their own inside `RefreshAsync` but forget to pass in the outer method's token.

Perhaps the .NET pattern for async cancellation is difficult in non-GCed languages, but it works well in .NET.